### ISSUE TITLE
Kiwix-serve 3.1.2-5, RPi, server error after a number of dice clicks + top-level searches 

### ISSUE THREAD

**Issue by eroyee** [kiwix/libkiwix#554]
After #551 I converted four 'old' .zim files using zimrecreate (without option) and trialled them. These were:

engineering.stackexchange 553.7MB
physics.stackexchange 3.6GB
wiktionary_en_all_maxi 6.1GB
latin.stackexchange 80.6MB

The PC is still chewing away with wikipedia_en_all_maxi, at this rate it'll be a few days.

In my previous tests I'd first used wikipedia, then random other libraries.

In this test I first used physics, then others.

I was not intially able to reproduce the error from #551, however when I did a TLD search across the four files (for 'coriolis' :-), then another for something else, the server failed with 'Internal Server Error ... Unexpected lzma status: 5'.

I've killed the process and tried this a couple of times, it's reproducible for me if I:

1. go into physics 
2. click the dice quite a few times 
3. then into say engineering 
4. click the dice several times 
5. then up to search across the four files for 'coriolis' (which works)
6. click into several returns
7. back up to search for something else (say 'loud')
8. then the issue occurs

Once it crashes the first time it's generally possible to go into individual files for a short time, after which it seems to crash more severely. On one occasion the file lost graphics/formatting, but text was still available.

Here's a couple of failure outputs from two different tests:

root@raspbx:/home/# ./kiwix-serve -d --port=8080 /mnt/usbdisk1/zims/engineering.stackexchange.com_en_all_2021-04a.zim /mnt/usbdisk1/zims/latin.stackexchange.com_en_all_2021-06a.zim /mnt/usbdisk1/zims/physics.stackexchange.com_en_all_2021-05a.zim /mnt/usbdisk1/zims/wiktionary_en_all_maxi_2021-06a.zim 
root@raspbx:/home/# Unexpected lzma status : 5
========== INTERNAL ERROR !! ============
Requesting : 
full_url : /search
method    : GET (0)
version   : HTTP/1.1
request#  : 429
headers   :
 - accept : 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8'
 - accept-encoding : 'gzip, deflate'
 - accept-language : 'en-US,en;q=0.5'
 - connection : 'keep-alive'
 - dnt : '1'
 - host : '192.168.1.50:8080'
 - referer : 'http://192.168.1.50:8080/'
 - upgrade-insecure-requests : '1'
 - user-agent : 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:87.0) Gecko/20100101 Firefox/87.0'
arguments :
 - pattern : 'coriolis'
Parsed : 
full_url: /search
url   : /search
acceptEncodingDeflate : 1
has_range : 0
is_valid_url : 1
.............
Response :
httpResponseCode : 500
headers :
 - Content-Type: 'text/html'
 - Access-Control-Allow-Origin: '*'
 - Cache-Control: 'no-cache, no-store, must-revalidate'
 - Content-Encoding: 'deflate'
 - Vary: 'Accept-Encoding'

---Sample from first test below, lots of similar output that filled up the buffer - likely because I kept on trying 'till it completely crashed ---

==== Unhandled error : Unexpected lzma status : 5
========== INTERNAL ERROR !! ============
Requesting :
full_url :
/engineering.stackexchange.com_en_all_2021-04a/A/question/33737.html
method    : GET (0)
version   : HTTP/1.1
request#  : 367
headers   :
  - accept :
'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8'
  - accept-encoding : 'gzip, deflate'
  - accept-language : 'en-US,en;q=0.5'
  - connection : 'keep-alive'
  - cookie :
'ys-epggrid=o%3Acolumns%3Da%253Ao%25253Aid%25253Ds%2525253Adetails%25255Ewidth%25253Dn%2525253A67%255Eo%25253Aid%25253Dn%2525253A1%25255Ewidth%25253Dn%2525253A100%255Eo%25253Aid%25253Ds%2525253Atitle%25255Ewidth%25253Dn%2525253A250%255Eo%25253Aid%25253Ds%2525253Asubtitle%25255Ewidth%25253Dn%2525253A250%255Eo%25253Aid%25253Ds%2525253AepisodeOnscreen%25255Ewidth%25253Dn%2525253A100%255Eo%25253Aid%25253Ds%2525253Astart%25255Ewidth%25253Dn%2525253A117%255Eo%25253Aid%25253Ds%2525253Astop%25255Ewidth%25253Dn%2525253A100%25255Ehidden%25253Db%2525253A1%255Eo%25253Aid%25253Ds%2525253Aduration%25255Ewidth%25253Dn%2525253A100%255Eo%25253Aid%25253Ds%2525253AchannelNumber%25255Ewidth%25253Dn%2525253A60%255Eo%25253Aid%25253Ds%2525253AchannelName%25255Ewidth%25253Dn%2525253A250%255Eo%25253Aid%25253Ds%2525253AstarRating%25255Ewidth%25253Dn%2525253A50%255Eo%25253Aid%25253Ds%2525253AageRating%25255Ewidth%25253Dn%2525253A50%255Eo%25253Aid%25253Ds%2525253Agenre%25255Ewidth%25253Dn%2525253A250%5Efilters%3Do%253A'
  - dnt : '1'
  - host : '192.168.1.50:8080'
  - referer :
'http://192.168.1.50:8080/engineering.stackexchange.com_en_all_2021-04a/A/index.html'
  - upgrade-insecure-requests : '1'
  - user-agent : 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:87.0)
Gecko/20100101 Firefox/87.0'
arguments :
Parsed :
full_url:
/engineering.stackexchange.com_en_all_2021-04a/A/question/33737.html
url   : /engineering.stackexchange.com_en_all_2021-04a/A/question/33737.html
acceptEncodingDeflate : 1
has_range : 0
is_valid_url : 1
.............
Response :
httpResponseCode : 500
headers :
  - Content-Type: 'text/html'
  - Access-Control-Allow-Origin: '*'
  - Cache-Control: 'no-cache, no-store, must-revalidate'
  - Content-Encoding: 'deflate'
  - Vary: 'Accept-Encoding'
===== Unhandled error : Unexpected lzma status : 5
===== Unhandled error : Broken promise
===== Unhandled error : Unexpected lzma status : 5

**Comment by eroyee** [comment]
Following up on this from some testing just now.

Having noted that the errors seemed to always return an 'lzma' message I recreated the files using zstd compression and removed the previous set.

At this point I've not been able to replicate the issue, which may narrow the focus a little. However I'd like to carry on some further testing before reaching a firm conclusion on this. 

I'm also in the process of setting up a faster machine to recreate the wikipedia .zim file (I had to abort the process) and would like to test with that. I expect this could take some days but will report back when it's done and I can trial it for while.

**Comment by mgautierfr** [comment]
Thank for the report. Hope than using zstd will fix you problem.

However, there is still a bug here. lzma should work correctly. If we have lzma error while decompressing we have to investigate this.

**Comment by eroyee** [comment]
@mgautierfr No worries, it's not a personal issue here - just trying to 
provide as much detail as I can in order to contribute meaningfully.

I'm not sure what implementation of lmza kiwix uses but I was interested 
to read this rather scathing article...:

http://www.nongnu.org/lzip/xz_inadequate.html

Also, my development machine has finished the wikipedia_en_all_maxi 
conversion - 82GB (presumably lzma?), 79.1GB zstd.

I've run that up in the Pi along with other zims and confirm that #551 
is sorted, and that [seemingly as long as the zim is created using zstd] 
#554 does not appear.

**Comment by kelson42** [comment]
Good to know the problem is fixed with zstd. Considering that lzma compression is deprecated And that we have been living with this big for years, I believe we should close this ticket as WONTFIX.

**Comment by eroyee** [comment]
A small suggestion if you do this.

Would it be useful to have zimrecreate (and maybe other tools?) default 
to zstd and require a switch to use lzma, rather than the other way 
around as it is presently?

**Comment by kelson42** [comment]
@eroyee yes, create a ticket. Wonder a bit why this is not the case already.

**Comment by kelson42** [comment]
@eroyee thx
